pennylane_lightning_gpu/src/simulator/StateVectorCudaMPI.hpp

```diff
@@ -941,6 +960,379 @@ class StateVectorCudaMPI
+                    splitSparseMatrix[block_row_id][block_col_id]
+                        .getCsrOffsets()[i0] +=
+                        splitSparseMatrix[block_row_id][block_col_id]
+                            .getCsrOffsets()[i0 - 1];
+                }
+            }
+        }
+
+        return splitSparseMatrix;
+    }
+
+    /**
+     * @brief expval(H) calculates the expected value using cuSparseSpMV and MPI
+     * to implement distributed Sparse Matrix-Vector multiplication. The dense
+     * vector is distributed across multiple GPU devices and only the MPI rank 0
+     * holds the complete sparse matrix data. The process involves the following
+     * steps: 1. The rank 0 splits the full sparse matrix into n by n blocks,
+     * where n is the size of MPI communicator. Each row of blocks is then
+     * distributed across multiple GPUs. 2. For each GPU, cuSparseSpMV is
+     * invoked to perform the local sparse matrix block and local state vector
+     * multiplication. 3. Each GPU will collect computation results for its
+     * respective row block of the sparse matrix. 4. After all sparse matrix
+     * operations are completed on each GPU, an inner product is performed and
+     * MPI reduce operation is ultilized to obtain the final result for the
+     * expectation value.
+     * @tparam index_type Integer type used as indices of the sparse matrix.
+     * @param csr_Offsets_ptr Pointer to the array of row offsets of the sparse
+     * matrix. Array of size csrOffsets_size.
+     * @param csrOffsets_size Number of Row offsets of the sparse matrix.
+     * @param columns_ptr Pointer to the array of column indices of the sparse
+     * matrix. Array of size numNNZ
+     * @param values_ptr Pointer to the array of the non-zero elements
+     * @param numNNZ Number of non-zero elements.
+     * @return auto Expectation value.
+     */
+    template <class index_type>
+    auto getExpectationValueOnSparseSpMV(
+        const index_type *csrOffsets_ptr, const index_type csrOffsets_size,
+        const index_type *columns_ptr,
+        const std::complex<Precision> *values_ptr, const index_type numNNZ) {
+        if (mpi_manager_.getRank() == 0) {
+            PL_ABORT_IF_NOT(static_cast<size_t>(csrOffsets_size - 1) ==
+                                (size_t{1} << this->getTotalNumQubits()),
+                            "Incorrect size of CSR Offsets.");
+            PL_ABORT_IF_NOT(numNNZ > 0, "Empty CSR matrix.");
+        }
+
+        const CFP_t alpha = {1.0, 0.0};
+        const CFP_t beta = {0.0, 0.0};
+
+        Precision local_expect = 0;
+
+        auto device_id = BaseType::getDataBuffer().getDevTag().getDeviceID();
+        auto stream_id = BaseType::getDataBuffer().getDevTag().getStreamID();
+
+        cudaDataType_t data_type;
+        cusparseIndexType_t compute_type;
+        cusparseOperation_t operation_type = CUSPARSE_OPERATION_NON_TRANSPOSE;
+        cusparseSpMVAlg_t spmvalg_type = CUSPARSE_SPMV_ALG_DEFAULT;
```

Comment:
We can likely make these `constexpr`, `const`, or `static const`, pending how much the compiler complains.

Reply:
Done!